### ISSUE TITLE
Missing dependency on a foreign library: zlib.h

### ISSUE THREAD

**Issue by gvolpe** [haskell/cabal#6228]
Hi, 

I'm getting this error while trying to compile a project:

```
Configuring library for zlib-0.6.2.1..
cabal: Missing dependency on a foreign library:
* Missing (or bad) header file: zlib.h
* Missing (or bad) C library: z
```

```
Build profile: -w ghc-8.6.5 -O1
```

I have seen people suggesting that installing the package `libghc-zlib-dev` fixes the issue. The problem is that I already have this package installed, however it's a newer version:

```
dpkg -l | grep zlib
ii  libghc-zlib-dev                            0.6.2-2                                 amd64        Compression and decompression in the gzip and zlib formats
ii  zlib1g:amd64                               1:1.2.11.dfsg-1ubuntu2                  amd64        compression library - runtime
ii  zlib1g:i386                                1:1.2.11.dfsg-1ubuntu2                  i386         compression library - runtime
ii  zlib1g-dev:amd64                           1:1.2.11.dfsg-1ubuntu2                  amd64        compression library - development
```

I'm on Ubuntu 19.04. Does anyone know how to fix this issue?

Thanks in advance,
Gabriel.

**Comment by hvr** [comment]
This is strange; after installing `zlib1g-dev` the error above should have been resolved. 

How did you install GHC? Was it via https://launchpad.net/~hvr/+archive/ubuntu/ghc or some other means?

Can you provide the verbose logging output (via `-v3`) for when cabal fails to locate zlib?

**Comment by gvolpe** [comment]
I installed both using `nix`:

```
$ which cabal
/home/gvolpe/.nix-profile/bin/cabal
$ which ghc
/home/gvolpe/.nix-profile/bin/ghc
```

Here's the output of `cabal new-build -v3`: https://pastebin.com/7midLWws

**Comment by fgaz** [comment]
What happens when you also install the `zlib` and `zlib.dev` nix packages?

---

A bit of a hack, but installing `'ghc.withPackages (hp: with hp; [ zlib ])'` instead of `ghc` should make it work in the meantime. I also had that problem so I have that line in my `configuration.nix`. For some reason other libraries work correctly.

**Comment by gvolpe** [comment]
I installed `zlib` using `nix-pkgs` too, but I get an error when trying to install `zlib.dev`. 

```
$ nix-env -i zlib
replacing old 'zlib-1.2.11'
installing 'zlib-1.2.11'
$ nix-env -i zlib.dev
error: selector 'zlib.dev' matches no derivations
```

I'm on the unstable channel:

```
$ nix-channel --list
nixpkgs https://nixos.org/channels/nixpkgs-unstable
```

I'm just getting started with `nixpkgs` in Ubuntu. Where do I find such configuration file?

**Comment by fgaz** [comment]
> I'm just getting started

Oh, sorry, I'll go into a bit more of detail then.

`nix-env -i` installs packages by name by searching for matching packages.

`nix-env -iA` installs packages by their attribute path, so for example `nixpkgs.zlib`. Other than being faster since it does not have to search, it can be used to specify subpackages or in general any attribute path in nixpkgs. So, to install `zlib.dev` you do `nix-env -iA nixpkgs.zlib.dev`

Finally, `ghc.withPackages (hp: with hp; [ zlib ])` is a (part of a) nix expression which tells nix to install ghc _and_ a global environment containing the zlib haskell package. You can install it by first removing the old ghc with `nix-env -e` and then running `nix-env -i -f '<nixpkgs>' -E 'pkgs: (pkgs {}).ghc.withPackages (hp: with hp; [ zlib ])'`.

> Where do I find such configuration file?

`configuration.nix` is only for NixOS, it is used to specify your system declaratively

**Comment by complyue** [comment]
I'm searching nix integration related issues and see this, I had the same problem, and I found simply add 
```nix
{
# ...
  # to workaround missing zlib.h problem during build
  librarySystemDepends = [ pkgs.zlib ];
}
```
to your `<pkg>.nix` (whether generated by cabal2nix or crafted by hand) can workaround it.

hope helpful to someone.

**Comment by CMCDragonkai** [comment]
How does this work when using `cabal2nix`. Is there a way to make the `package.yaml` specify the correct zlib to get.

**Comment by CMCDragonkai** [comment]
Posted here too: https://github.com/NixOS/cabal2nix/issues/437

**Comment by chrissound** [comment]
This should probably be closed as it's a nix related issues (that seems to be solved in the link thread).

**Comment by iowaguy** [comment]
For posterity, the solution referenced above from the Nix manual is no longer... in the manual. I did some sleuthing and found it in [the internet archive](https://web.archive.org/web/20191214012041/https://nixos.org/nixpkgs/manual/) (see section 15.9.4.7. Builds using Stack complain about missing system libraries).

In short, just add the following to `stack.yml`:

```
nix:
  enable: true
  packages: [ zlib ]
```

**Comment by GavinRay97** [comment]
In case anyone lands here from a Google search due to below error (like I did) and is on a Debian-based distro (IE, Ubuntu)

```sh
Warning: Failed to build documentation for basement-0.0.11 (which is required
by exe:mylib from mylib-0.1.0.0).
```

There are 2 things I've found:

1. You need `zlib` installed in the system
    - `prettyprinter` and `basement` failed to build without this
2. You need to set locale (`export LC_ALL=C.UTF-8`) for `basement` docs to build, see Debian bug and patch here:
    - https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=981718
    - https://www.mail-archive.com/pkg-haskell-maintainers@alioth-lists.debian.net/msg16230.html

```diff
diff -Nru haskell-basement-0.0.11/debian/changelog haskell-basement-0.0.11/debian/changelog
--- haskell-basement-0.0.11/debian/changelog	2020-05-26 22:07:56.000000000 +0530
+++ haskell-basement-0.0.11/debian/changelog	2021-02-17 21:48:15.000000000 +0530
@@ -1,3 +1,10 @@
+haskell-basement (0.0.11-1.1) unstable; urgency=medium
+
+  * Non-maintainer upload.
+  * Explicitly set LANG to utf8 to workaround a build bug (Closes: #981718) 
+
+ -- Ritesh Raj Sarraf <rrs@debian.org>  Wed, 17 Feb 2021 21:48:15 +0530
+
 haskell-basement (0.0.11-1) unstable; urgency=medium
```
```diff
diff -Nru haskell-basement-0.0.11/debian/rules haskell-basement-0.0.11/debian/rules
--- haskell-basement-0.0.11/debian/rules	2020-02-08 07:28:07.000000000 +0530
+++ haskell-basement-0.0.11/debian/rules	2021-02-17 21:48:11.000000000 +0530
@@ -4,5 +4,6 @@
 DEB_CABAL_PACKAGE = basement
 DEB_DEFAULT_COMPILER = ghc
 
+export LC_ALL=C.UTF-8
 include /usr/share/cdbs/1/rules/debhelper.mk
 include /usr/share/cdbs/1/class/hlibrary.mk
```
